include/tvm/graph/internal/DependencyGraph.h

```diff
@@ -0,0 +1,170 @@
+        * where the rank is a heuristic number approximating the tree depth. The
+        * root with the smallest rank point to the other, in which case the rank of
+        * the other remain unchanged. If both roots have the same rank, the second
+        * point to the first and the rank of the first one is incremented by one.
+        * The ranks are stored in the aptly named \p rank.
+        */
+      void unify(size_t node1, size_t node2);
+
+      /** Return the parent of node i*/
+      bool isRoot(size_t i) const;
+
+    private:
+      /** Find the root of the tree the element \p node is in.
+        * The function also perform a path compression, in that all node discovered
+        * along the way are made to point directly to the root.
+        */
+      size_t rootCompr(size_t node);
+
+      std::vector<size_t> parent_;  // parent_[i] is the id of the parent of element i
+      std::vector<size_t> rank_;    // rank of each element.
+    };
+
+    /** A recursive function that finds and prints strongly connected
+      * components (SCC) using DFS traversal
+      * \param ret a vector of SCC
+      * \param u the vertex to visit
+      * \param disc discovery time of the vertices (should be -1 if not
+      * discovered yet)
+      * \param low earliest visited vertex (the vertex with minimum discovery
+      * time) that can be reached from subtree rooted with current vertex
+      * \param st stack to store all connected ancestors
+      * \param stackMember array for checking faster if a node is in the stack
+      * \param time a counter increased at each entry in the function. Should
+      * be non-negative.
+      *
+      * Adapted from
+      * https://www.geeksforgeeks.org/tarjan-algorithm-find-strongly-connected-components/
+      */
+    void SCCUtil(std::vector<std::vector<size_t>>& ret,
+      size_t u, std::vector<int>& disc, std::vector<int>& low,
+      std::stack<size_t>& st, std::vector<bool>& stackMember, int& time) const;
+
+    /** An intermediate function for computing the topological order.*/
+    std::pair<std::vector<size_t>, DisjointSet> order_() const;
+
+
+    /** A recursive function used for topological ordering. It also detects
+      * connected components of the graph.
+      * \param v the vertex to process
+      * \param order the vertices in reversed topological order
+      * \param visited visited[i] is true if and only if vertex i was visited
+      * \param stack stack[i] is true if and only if vertex i is among the
+      * vertices being processed.
+      * \param components A disjoint-set data structure representing the
+      * connected components
+      */
+    void orderUtil(size_t v, std::vector<size_t>& order,
+      std::vector<bool>& visited, std::vector<bool>& stack,
+      DisjointSet& components) const;
+
+
+    std::vector<bool> roots_;                       //roots_[i] is true iff the i-th node has no incoming edges
```

Comment:
What would you recommend here? `std::vector<char>`/`std::vector<uint8>` ?

I just passed some time reading about `std::vector<bool>` (supposed) problems. I get to two main conclusions:
 * `std::vector<bool>` can break code or behave in surprising ways because it does not meet the standard’s container or sequence requirements, with in particular problems around reference on an element (see [here](https://riptutorial.com/cplusplus/example/2561/vector-bool---the-exception-to-so-many--so-many-rules) or [here](https://riptutorial.com/cplusplus/example/2561/vector-bool---the-exception-to-so-many--so-many-rules) )
 * it's rather inconclusive about the performance issue, and does depend on the use case (see e.g. [here](https://isocpp.org/blog/2012/11/on-vectorbool), [here](https://stackoverflow.com/questions/36932684/c11-vectorbool-performance-issue-with-code-example) or [here](https://codeforces.com/blog/entry/57052))

So ideally, we should benchmark before making a change. I'm not sure it is worth the time or the more pressing possible optimization in TVM.

Reply:
> So ideally, we should benchmark before making a change. I'm not sure it is worth the time or the more pressing possible optimization in TVM.

Indeed, it's probably not measurable in the grand scheme of things but the only two operations we perform on this vector are:
- `push_back`
- random access (`operator[]`)

I wrote a short benchmark based on this and the posts you linked (source [here](https://gist.github.com/gergondet/8a736482500329c73f5e644d76819a52))

```
--------------------------------------------------------------------
Benchmark                          Time             CPU   Iterations
--------------------------------------------------------------------
BM_vector_access<bool>          2686 ns         2686 ns       229716
BM_vector_access<uint8_t>       1675 ns         1675 ns       414418
BM_vector_access<int8_t>        1059 ns         1059 ns       655812
BM_vector_push<bool>            1354 ns         1354 ns       511126
BM_vector_push<uint8_t>          272 ns          272 ns      2554368
BM_vector_push<int8_t>           536 ns          536 ns      1275125
```

(note: the results are not significative for the int8_t/uint8_t difference, they swap around if I swap the benchmark order)

So I think `std::vector<uint8_t>` is a fitting replacement (and it has zero impact on the code you write)

Reply:
Done in fcdd4ebd8b2e60b55d8260b3da180f91989ae42b